pkg/ottl/ottlfuncs/func_scale.go

```diff
@@ -0,0 +1,137 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ottlfuncs // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+type ScaleArguments[K any] struct {
+	Value      ottl.Getter[K]
+	Multiplier float64
+}
+
+func NewScaleFactory[K any]() ottl.Factory[K] {
+	return ottl.NewFactory("Scale", &ScaleArguments[K]{}, createScaleFunction[K])
+}
+
+func createScaleFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[K], error) {
+	args, ok := oArgs.(*ScaleArguments[K])
+
+	if !ok {
+		return nil, fmt.Errorf("ScaleFactory args must be of type *ScaleArguments[K]")
+	}
+
+	return Scale(args.Value, args.Multiplier)
+}
+
+func Scale[K any](getter ottl.Getter[K], multiplier float64) (ottl.ExprFunc[K], error) {
+	return func(ctx context.Context, tCtx K) (any, error) {
+		got, err := getter.Get(ctx, tCtx)
+		if err != nil {
+			return nil, err
+		}
+
+		switch value := got.(type) {
+		case float64:
+			return value * multiplier, nil
+		case int64:
+			return float64(value) * multiplier, nil
+		case pmetric.NumberDataPointSlice:
+			scaledMetric := pmetric.NewNumberDataPointSlice()
+			value.CopyTo(scaledMetric)
+			scaleMetric(scaledMetric, multiplier)
+			return scaledMetric, nil
```

Comment:
I know in the issue we were trending toward `Converter`, but now looking at the implementation the inefficiency of copying the entire metric to scale it, returning the copy, and then depending on set to copy it back feels so inefficient.

I think a scale_metric editor would be better here.

Reply:
alright, I will adapt the function to be an editor instead. will ping you once I have updated the PR

Reply:
@TylerHelmuth Regarding the naming: Do you think it would be better to just call it `scale`, instead of `scale_metric`? The latter suggests that this will only work for metrics, while it could potentially be used for numeric attributes as well

Reply:
I think we can probably forgo the need to scale numeric values with this function, since we already have arithmetic in OTTL that should work for each data point value type. If we allow this function to handle data points we gain the ability to scale individual histogram data points, but that's probably an edge case compared to the common case of scaling all data points for a particular metric.

Reply:
good point - I have adjusted the PR to only support metrics now - I also noticed that in the e2e tests only log transform contexts were used for the test cases, so i added a new test using a metric transform context and equivalent helper functions. that should hopefully proof to be useful going forward for other metric-specific functions